auditlog: add documentation (ru,en) (#2899)
ydb-platform/ydb: ydb/docs/en/core/security/audit-log.md

# Audit logging

> _Audit log_ is a stream which records user induced activity with objects and data in {{ ydb-short-name }}.

Reviewer: ```suggestion
The _audit log_ is a stream that records user-induced activity with objects and data in {{ ydb-short-name }}.
```

> Audit log makes it possible to monitor activities with [schema entities](audit-log-scheme.md), [data in tables](audit-log-dml.md), and track [client connections](audit-log-conn.md). Basically, audit log provides information about what actions were performed on which objects from which ip addresses on behalf of what users or services. As such, audit log helps organizations to create a historical record of activity to comply with regulatory requirements or enforce other policies.

Reviewer: ```suggestion
The audit log makes it possible to monitor activities with [schema entities](audit-log-scheme.md), [data in tables](audit-log-dml.md), and track [client connections](audit-log-conn.md). Essentially, the audit log provides information about what actions were performed on which objects, from which IP addresses, and on behalf of which users or services. As such, the audit log helps organizations create a historical activity record to comply with regulatory requirements or enforce other policies.
```

## Audit log records and attributes

> Information about actions or operations is recorded in the audit log in the form of records. Each record consists of a set of fields or attributes. Some attributes are the same for all records, others are determined by the type of activity or the type of operation described.

Reviewer: ```suggestion
Information about actions or operations is recorded in the audit log in the form of records. Each record consists of a set of fields or attributes. Some attributes are the same for all records, while others are determined by the type of activity or the type of operation described.
```

> | __Attribute__ | __Description__ |
> |:----|:----|
> | **Common** |
> | `component` | Name of the {{ ydb-short-name }} component that generated the event (for example, `schemeshard`, `grpc-conn`, `grpc-proxy`). |
> | `subject` | User SID (`<login>@<subsystem>` format). Unless mandatory authentication is enabled, the attribute will be set to `{none}`.<br>Required. |
> | `operation` | Names of operations or actions are similar to the YQL syntax (for example, `ALTER DATABASE`, `CREATE TABLE`).<br>Required. |
> | `database` | Database path (for example, `/root/db`). |
> | `status` | General operation status.<br>Possible values:<ul><li>`SUCCESS`: The operation completed successfully.</li><li>`ERROR`: The operation failed.</li></ul>Required. |
> | `detailed_status` | Detailed operation status. Component specific. |
> | `reason` | Error message. |
> | `request_id` | Unique ID of the request that invoked the operation. You can use the `request_id` to differentiate events related to different operations and link the events together to build a single audit-related operation context. |
> | `remote_address` | The IP of the client that delivered the request. |
> | `start_time` | Operation start time. |
> | `end_time` | Operation end time. |
> | **Schema operations** | (`schemeshard` component)
> | `tx_id` | Unique transaction ID. Similarly to `request_id`, this ID can be used to differentiate events related to different operations.<br>Required. |
> | `paths` | List of paths in the database that are changed by the operation (for example, `[/root/db/table-a, /root/db/table-b]`).<br>Required. |
> | _Ownership and permissions_ ||
> | `new_owner` | User SID of the new owner of the object when ownership is transferred.  |
> | `acl_add` | List of added permissions, in [short notation](./short-access-control-notation.md) (for example, `[+R:someuser]`). |
> | `acl_remove` | List of revoked permissions, in [short notation](./short-access-control-notation.md) (for example, `[-R:somegroup]`). |
> | _[User (or custom) attributes](../dev/custom-attributes.md)_ |
> | `user_attrs_add` | List of custom attributes added when creating objects or updating attributes (for example, `[attr_name1: A, attr_name2: B]`). |
> | `user_attrs_remove` | List of custom attributes removed when creating objects or updating attributes (for example, `[attr_name1, attr_name2]`). |
> | **[Credentials](../concepts/auth#static-credentials) management operations** | (`schemeshard` component)
> | `login_user` | User name (when [adding or modifying the user](./access-management.md#users)).
> | `login_group` | Group name (when [adding or modifying the user group](./access-management.md#groups)).
> | `login_member` | User name (when [adding or removing the user to or from a group](./access-management.md#groups)).
> | **DML operations** | (`grpc-proxy` component)
> | `query_text` | YQL query text.
> | `prepared_query_id` | Prepared query ID.
> | `tx_id` | Unique transaction ID.
> | `begin_tx` | Request for implicit transaction start.
> | `commit_tx` | Request for implicit transaction commit.
> | `table` | Path of the affected table.
> | `row_count` | Number of the changed rows.

Reviewer: ```suggestion
| __Attribute__ | __Description__ |
|:----|:----|
| **Common** |
| `component` | Name of the {{ ydb-short-name }} component that generated the event (for example, `schemeshard`, `grpc-conn`, `grpc-proxy`). |
| `subject` | User SID (`<login>@<subsystem>` format). Unless mandatory authentication is enabled, the attribute will be set to `{none}`.<br>Required. |
| `operation` | Names of operations or actions similar to the YQL syntax (for example, `ALTER DATABASE`, `CREATE TABLE`).<br>Required. |
| `database` | Database path (for example, `/root/db`). |
| `status` | General operation status.<br>Possible values:<ul><li>`SUCCESS`: The operation completed successfully.</li><li>`ERROR`: The operation failed.</li></ul>Required. |
| `detailed_status` | Detailed operation status, component-specific. |
| `reason` | Error message. |
| `request_id` | Unique ID of the request that invoked the operation. You can use the `request_id` to differentiate events related to different operations and link the events together to build a single audit-related operation context. |
| `remote_address` | The IP address of the client that delivered the request. |
| `start_time` | Operation start time. |
| `end_time` | Operation end time. |
| **Schema operations** | (`schemeshard` component) |
| `tx_id` | Unique transaction ID. Similar to `request_id`, this ID can be used to differentiate events related to different operations.<br>Required. |
| `paths` | List of paths in the database that are changed by the operation (for example, `[/root/db/table-a, /root/db/table-b]`).<br>Required. |
| _Ownership and permissions_ ||
| `new_owner` | User SID of the new owner of the object when ownership is transferred. |
| `acl_add` | List of added permissions, in [short notation](./short-access-control-notation.md) (for example, `[+R:someuser]`). |
| `acl_remove` | List of revoked permissions, in [short notation](./short-access-control-notation.md) (for example, `[-R:somegroup]`). |
| _[User (or custom) attributes](../dev/custom-attributes.md)_ ||
| `user_attrs_add` | List of custom attributes added when creating objects or updating attributes (for example, `[attr_name1: A, attr_name2: B]`). |
| `user_attrs_remove` | List of custom attributes removed when creating objects or updating attributes (for example, `[attr_name1, attr_name2]`). |
| **[Credentials](../concepts/auth#static-credentials) management operations** | (`schemeshard` component) |
| `login_user` | User name (when [adding or modifying the user](./access-management.md#users)). |
| `login_group` | Group name (when [adding or modifying the user group](./access-management.md#groups)). |
| `login_member` | User name (when [adding or removing the user to or from a group](./access-management.md#groups)). |
| **DML operations** | (`grpc-proxy` component) |
| `query_text` | YQL query text. |
| `prepared_query_id` | Prepared query ID. |
| `tx_id` | Unique transaction ID. |
| `begin_tx` | Request for implicit transaction start. |
| `commit_tx` | Request for implicit transaction commit. |
| `table` | Path of the affected table. |
| `row_count` | Number of changed rows. |
```

## Usage and configuration

Audit logging subsystem supports recording the stream to:

> * a file on each {{ ydb-short-name }} cluster node;
> * standard error stream `stderr`;
> * [Unified Agent](https://cloud.yandex.com/en/docs/monitoring/concepts/data-collection/unified-agent/) &mdash; agent for delivering metrics and logs.

Reviewer: ```suggestion
* A file on each {{ ydb-short-name }} cluster node.
* Standard error stream (`stderr`).
* [Unified Agent](https://cloud.yandex.com/en/docs/monitoring/concepts/data-collection/unified-agent/), an agent for delivering metrics and logs.
```

> It is possible to use any of the listed destinations, as well as their combination.

Reviewer: ```suggestion
It is possible to use any of the listed destinations or their combination.
```

> Writing to a file or unified-agent is recommended for use in production installations. In the case of writing to a file, access to the audit log is set by file-system rights.

Reviewer: ```suggestion
Writing to a file or using Unified Agent is recommended for production deployments. In the case of writing to a file, access to the audit log is controlled by file system permissions.
```

> Writing to `stderr` is recommended for test installations. Further processing of `stderr` may be affected by the settings of the [logging](../devops/manual/logging.md) subsystem.

Reviewer: ```suggestion
Writing to `stderr` is recommended for test installations. The settings of the [logging](../devops/manual/logging.md) subsystem may influence its further processing.
```

### Settings

> Audit logging require configuration at the {{ ydb-short-name }} cluster level.

Reviewer: ```suggestion
Audit logging requires configuration at the {{ ydb-short-name }} cluster level.
```

> It should be configured by adding `audit_config` section with a set of destination backends to the [cluster configuration](../deploy/configuration/config.md):

Reviewer: ```suggestion
It should be configured by adding the `audit_config` section with a set of destination backends to the [cluster configuration](../deploy/configuration/config.md):
```

```yaml
audit_config:
  file_backend:
    format: OUTPUT_FORMAT
    file_path: "path_to_file"
  unified_agent_backend:
    format: OUTPUT_FORMAT
    log_name: session_meta_log_name
  stderr_backend:
    format: OUTPUT_FORMAT
```

The`audit_config` section should not be empty.

All backends (`file_backend`, `unified_agent_backend`, `stderr_backend`) are optional.

| Key | Description |
--- | ---

> | `*.format` | Audit log format. <br>Acceptable values:<ul><li>`JSON`: Serialized [JSON]{% if lang == "ru" %}(https://ru.wikipedia.org/wiki/JSON){% endif %}{% if lang == "en" %}(https://en.wikipedia.org/wiki/JSON){% endif %}.</li><li>`TXT`: Text format.</ul>_Optional_.<br>Default value: `JSON`.
> | `file_backend` | Local file backend.
> | `file_backend.file_path` | Path to a file. Any path component that is missing will be created on each node at cluster startup. If the file exists, the data will be appended to it.<br>_Required for `file_backend`_.
> | `unified_agent_backend` | Unified Agent backend.<br>Unified Agent client should be configured by `uaclient_config` section in the [cluster configuration](../deploy/configuration/config.md).
> | `unified_agent_backend.log_name` | A field in the Unified Agent session metadata that will be passed along with the messages. Allows log stream dispatch to one or more child channels according to the condition `_log_name: "session_meta_log_name"`.<br>_Optional_.<br>Default value: value of `uaclient_config.log_name`.
> | `stderr_backend` | `stderr` backend.
> 
> ### How to enable {#enabling-audit-log}
> 
> Audit logging is disabled by default.
> 
> Adding `audit_config` section to the [cluster configuration](../deploy/configuration/config.md) is enough to enable audit logging.
> 
> Additional requirements for enabling logging of individual activities are described in the relevant sections.

Reviewer: ```suggestion
| `*.format` | Audit log format. <br>Acceptable values:<ul><li>`JSON`: Serialized [JSON](https://en.wikipedia.org/wiki/JSON).</li><li>`TXT`: Text format.</ul>_Optional_.<br>Default value: `JSON`. |
| `file_backend` | Local file backend. |
| `file_backend.file_path` | Path to a file. Any missing path component will be created on each node at cluster startup. If the file exists, the data will be appended to it.<br>_Required for `file_backend`_. |
| `unified_agent_backend` | Unified Agent backend.<br>Unified Agent client should be configured by the `uaclient_config` section in the [cluster configuration](../deploy/configuration/config.md). |
| `unified_agent_backend.log_name` | A field in the Unified Agent session metadata that will be passed along with the messages. Allows log stream dispatch to one or more child channels according to the condition `_log_name: "session_meta_log_name"`.<br>_Optional_.<br>Default value: value of `uaclient_config.log_name`. |
| `stderr_backend` | `stderr` backend. |

### How to enable {#enabling-audit-log}

Audit logging is disabled by default.

Enabling audit logging requires adding the `audit_config` section to the [cluster configuration](../deploy/configuration/config.md).

Additional requirements for enabling logging of individual activities are described in the relevant sections.
```
